### PR TITLE
Add LLDP receive support

### PR DESCRIPTION
**Pull request by olbjo**
Draft implementation of LLDP receive feature
Alarms, timers and integration with rest of stack is not yet implemented. Please check overall structure and parsing strategy.

- Based on contributions by atomic-3dpenguin
- Adds parsing of incoming LLDP packets

### CODE REVIEW

File: src/common/pf_lldp.c [hunk 1]
```diff
@@ -440,10 +527,167 @@ void pf_lldp_send(
...
+         lldp_peer_cfg->ttl = pf_get_uint16(&parse_info, &offset);
+         /*
+          * TODO - Configure timeout
+          */
+         break;
+      case LLDP_TYPE_ORG_SPEC:
+         {
+            uint8_t org_id[3];
+            uint8_t org_subtype;
+            pf_get_mem(&parse_info, &offset, 3, org_id);
```
Comment (pyhys) [inline src/common/pf_lldp.c:637]: Use sizeof() instead of 3? (multiple places)

File: src/common/pf_lldp.h [hunk 2]
```diff
@@ -43,6 +43,19 @@ void pf_lldp_init(
 void pf_lldp_send(
    pnet_t                  *net);
 
+/**
+ * Receive an LLDP message.
+ *
+ * Parse LLDP tlv format and store selected information.
+ * @param net              InOut: The p-net stack instance
+ * @param p_buf            In:    The Ethernet frame.
+ * @param frame_id_pos     In:    The position of the frame id.
```
Comment (pyhys) [inline src/common/pf_lldp.h:52]: The parameter is named frame_pos, offset and frame_id_pos here and in the .c file. Make it consistent. Make sure it is the position of the frame id that actually is inserted.
Also p_buf is inconsistently named.

File: src/common/pf_lldp.c [hunk 3]
```diff
@@ -440,10 +527,167 @@ void pf_lldp_send(
...
+      case LLDP_TYPE_PORT_ID:
+         if ((frame.len > 0) &&
+             (frame.len - 1) < sizeof(lldp_peer_cfg->port_id))
+         {
+            lldp_peer_cfg->port_id_len = frame.len - 1;
+            lldp_peer_cfg->port_id_subtype = pf_get_byte(&parse_info, &offset);
+            pf_get_mem(&parse_info, &offset, lldp_peer_cfg->port_id_len, &lldp_peer_cfg->port_id);
+            lldp_peer_cfg->chassis_id[lldp_peer_cfg->port_id_len] = '\0';
+
+            /* Update Alias name as follows:
```
Comment (pyhys) [inline src/common/pf_lldp.c:591]: Implement a function to generate the alias name.
```
/**
 * @internal
 * Generate an alias name from port_id and chassis_id
 *
 * Defined in Profinet 2.4, section 4.3.1.4.18
 *
 * Check if the LLDP_TYPE_PORT_ID contains a "." (Example: port-001.test)
 *  If it does copy this over to the alias name
 *  If no "." is found concatenate the port id and chassis id
 *
 * @param port_id          In:   Peer port ID. Terminated string.
 * @param chassis_id       In:   Peer chassis ID. Terminated string.
 * @param alias            Out:  Alias name. Terminated string.
 * @param len              In:   Size of output buffer
 * @return  0  if the operation succeeded.
 *          -1 if an error occurred.
 */
int pf_lldp_generate_alias_name(
const char*             port_id,
const char*             chassis_id,
char*             	alias,
uint16_t                len)
{
}
```


Implement unit tests. Use: 
`class LldpUnitTest : public PnetUnitTest {};`

See for example:
`pf_cmina_is_stationname_valid()`

File: src/common/pf_lldp.c [hunk 4]
```diff
@@ -20,17 +20,18 @@
...
  * ToDo: Handle PNET_MAX_PORT ports.
- * ToDo: Receive LLDP and build a per-port peer DB.
  */
 
 #include <string.h>
-
+#include <ctype.h>
 #include "pf_includes.h"
 #include "pf_block_writer.h"
+#include "pf_block_reader.h"
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:34]: ok

File: src/common/pf_lldp.c [hunk 5]
```diff
@@ -54,23 +69,64 @@ typedef enum lldp_pnio_subtype_values
...
    { 0x01, 0x80, 0xc2, 0x00, 0x00, 0x0e }       /* LLDP Multicast */
 };
 
+/*
+ * TODO - remove before merge
+ */
+#undef LLPD_DEBUG
+#ifdef LLPD_DEBUG
+#define LLDP_DUMP_BUF(p, len) dump_buf(p,len)
+static void dump_buf (const uint8_t *p, uint16_t len)
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:89]: ok

File: src/common/pf_lldp.c [hunk 6]
```diff
@@ -79,15 +135,50 @@ static const pnet_ethaddr_t   lldp_dst_addr = {
...
+ * Parse LLDP TLV header.
+ * Each TVL is structured as follows:
+ * - Type   = 7 bits
+ * - Length = 9 bits
+ * - data   = 0-511 bytes
+ *
+ * @param p_info           InOut: Parser state.
+ * @param p_pos            InOut: Offset in parsed buffer
+ */
+lldp_frame_t pf_get_lldp_tlv_header(
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:159]: ok

File: src/common/pf_lldp.c (hunk 3 above)
Comment (olbjo) [reply src/common/pf_lldp.c]: Ok,
I split the receive operation in a parser and an action function.

File: src/common/pf_lldp.c [hunk 7]
```diff
@@ -54,23 +72,33 @@ typedef enum lldp_pnio_subtype_values
...
    LLDP_PNIO_SUBTYPE_MRP_INTERCONNECTION_PORT_STATUS = 8,
    /* 0x09..0xff reserved */
 } lldp_pnio_subtype_values_t;
 
-static const pnet_ethaddr_t   lldp_dst_addr = {
+static const pnet_ethaddr_t lldp_dst_addr = {
    { 0x01, 0x80, 0xc2, 0x00, 0x00, 0x0e }       /* LLDP Multicast */
 };
 
+#define LLPD_DEBUG
```
Comment (pyhys) [inline src/common/pf_lldp.c:86]: Remember to remove this

File: src/common/pf_lldp.c [hunk 8]
```diff
@@ -440,10 +533,239 @@ void pf_lldp_send(
...
+}
+
+/**
+ * @internal
+ * Generate an alias name from port_id and chassis_id
+ *
+ * TODO - It is uncertain if the alias is name generation
+ *        should depend on the port and chassis id sub types.
+ *        Current implementation may work but uncertain why.
+ *
```
Comment (pyhys) [inline src/common/pf_lldp.c:675]: Defined in Profinet 2.4, protocol section 4.3.1.4.18

File: src/common/pf_lldp.c [hunk 9]
```diff
@@ -440,10 +533,239 @@ void pf_lldp_send(
...
    {
       LOG_ERROR(PF_ETH_LOG, "LLDP(%d): Failed to reschedule LLDP sending\n", __LINE__);
    }
 }
+
+/**
+ * @internal
+ * Parse LLDP frame buffer
+ *
+ * @param buf              :In Reference to LLDP data
```
Comment (pyhys) [inline src/common/pf_lldp.c:550]: The stack uses this convention:
@param port_id          In:   Peer port ID. Terminated string.

File: src/common/pf_lldp.h [hunk 10]
```diff
@@ -43,6 +44,34 @@ void pf_lldp_init(
...
+
+/************ Internal functions, made available for unit testing ************/
+
+int pf_lldp_generate_alias_name(
+   const char*             port_id,
+   const char*             chassis_id,
+   char*                   alias,
+   uint16_t                len);
+
+int pf_lldp_parse_packet(
```
Comment (pyhys) [inline src/common/pf_lldp.h:70]: We can skip this one for the moment, as there are no tests defined.

File: test/test_lldp_unit.cpp [hunk 11]
```diff
@@ -0,0 +1,55 @@
+/*********************************************************************
```
Comment (pyhys) [inline test/test_lldp_unit.cpp:1]: I think we can put this in the existing test/test_lldp.cpp file

**Review by pyhys — APPROVED** [review]
Great! Please convert it to non-draft, and let HE have a look after the minor changes suggested.

File: src/common/pf_lldp.c (hunk 7 above)
Comment (olbjo) [reply src/common/pf_lldp.c:86]: Good point

File: src/common/pf_lldp.c (hunk 8 above)
Comment (olbjo) [reply src/common/pf_lldp.c:675]: Ok

File: src/common/pf_lldp.c (hunk 9 above)
Comment (olbjo) [reply src/common/pf_lldp.c:550]: OK

File: test/test_lldp_unit.cpp (hunk 11 above)
Comment (olbjo) [reply test/test_lldp_unit.cpp:1]: ok

**Comment by olbjo** [comment]
When running test on another feature I detected problems with the LLDP implementation so set pullrequest back to draft

File: include/pnet_api.h [hunk 12]
```diff
@@ -1027,17 +1035,78 @@ typedef struct pnet_ethaddr
...
+typedef struct pnet_lldp_peer_to_peer_boundary
+{
+   pnet_lldp_boundary_t boundary;
+   uint16_t properites;
+} pnet_lldp_peer_to_peer_boundary_t;
+
+/**
+ * LLDP Peer information used by the Profinet stack.
+ */
+typedef struct pnet_lldp_peer_cfg
```
Comment (fredrikdanielmoller) [inline include/pnet_api.h:1087]:
```suggestion
typedef struct pnet_lldp_peer_info
```

Comment (olbjo) [reply include/pnet_api.h:1087]: ok

File: test/test_lldp.cpp [hunk 13]
```diff
@@ -59,3 +59,28 @@ TEST_F(LldpTest, LldpInitTest)
    EXPECT_EQ(appdata.call_counters.write_calls, 0);
    EXPECT_EQ(mock_os_data.eth_send_count, 1);
 }
+
+TEST_F(LldpTest, LldpGenerateAliasName)
```
[the remark this answers is not in the document]
Comment (olbjo) [reply test/test_lldp.cpp:63]: Ok, I have added some test cases

File: src/common/pf_lldp.c [hunk 14]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+         }
+         break;
+      case LLDP_TYPE_PORT_ID:
+         if ((frame.len > 0) &&
+             ((size_t)(frame.len - 1) < sizeof(lldp_peer_info->port_id)))
+         {
+            lldp_peer_info->port_id_len = frame.len - 1;
+            lldp_peer_info->port_id_subtype = pf_get_byte(&parse_info, &offset);
+            pf_get_mem(&parse_info, &offset, lldp_peer_info->port_id_len, &lldp_peer_info->port_id);
+            lldp_peer_info->chassis_id[lldp_peer_info->port_id_len] = '\0';
```
Comment (fredrikdanielmoller) [inline src/common/pf_lldp.c:590]:
```suggestion
            lldp_peer_info->port_id[lldp_peer_info->port_id_len] = '\0';
```

File: src/common/pf_lldp.c [hunk 15]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+         }
+         else
+         {
+            LOG_ERROR(PF_LLDP_LOG, "LLDP(%d): Port Id too long (%d)\n", __LINE__, frame.len - 1);
```
Comment (fredrikdanielmoller) [inline src/common/pf_lldp.c:594]: Update 'offset' variable?

File: src/common/pf_lldp.c [hunk 16]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+               case LLDP_IEEE_SUBTYPE_MAC_PHY:
+                  lldp_peer_info->phy_config.aneg_support_status   = pf_get_byte(&parse_info, &offset);
+                  lldp_peer_info->phy_config.aneg_cap              = pf_get_uint16(&parse_info, &offset);
+                  lldp_peer_info->phy_config.operational_mau_type  = pf_get_uint16(&parse_info, &offset);
+                  break;
+               default:
+                  offset += (frame.len - 4);
+                  break;
+               }
+            }
```
Comment (fredrikdanielmoller) [inline src/common/pf_lldp.c:642]: In the else-case, should 'offset' be updated?

**Review by fredrikdanielmoller — CHANGES_REQUESTED** [review]

File: src/common/pf_lldp.c [hunk 17]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+   if (((strlen(port_id) + 1) > len) ||
+       ((strlen(port_id) + strlen(chassis_id) + 2 ) > len))
+   {
+      return -1;
+   }
+
+   memset(alias, '\0', len);
+   if (strchr(port_id,'.') != NULL)
+   {
+      strncpy(alias, port_id, len);
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:695]: Ok, memset removed. Updated the conditions for checking the string lengths.

File: src/common/pf_lldp.c [hunk 18]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+ * @internal
+ * Parse LLDP frame buffer
+ *
+ * @param buf              In:    LLDP data.
+ * @param len              In:    Length of LLDP data.
+ * @param lldp_peer_info   InOut: Record holding parsed LLDP information.
+ * @return 0 if parsing is successful, -1 on error.
+ */
+int pf_lldp_parse_packet(
+      uint8_t               buf[],
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:548]: ok

File: src/common/pf_lldp.c [hunk 19]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+       * TODO - pf_lldp_send_remote_mismatch_alarm(net);
+       */
+   }
+
+   memcpy(stored_peer_info, lldp_peer_info, sizeof(pnet_lldp_peer_info_t));
+}
+
+void pf_lldp_recv(
+    pnet_t              *net,
+    os_buf_t            *p_frame_buf,
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:746]: I agree, but this is aligned with pf_eth_recv(). Other reused functions used for parsing needs to be updated if I change this so I would prefer to skip this. Is it ok to skip update?

[the remark this answers is not in the document]
Comment (fredrikdanielmoller) [reply src/common/pf_lldp.c:746]: It's ok to skip. A separate pull request could add the 'const'.

File: src/common/pf_lldp.c [hunk 20]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+             ((size_t)(frame.len - 1) < sizeof(lldp_peer_info->chassis_id)))
+         {
+            lldp_peer_info->chassis_id_len = frame.len - 1;
+            lldp_peer_info->chassis_id_subtype = pf_get_byte(&parse_info, &offset);
+            pf_get_mem(&parse_info, &offset, lldp_peer_info->chassis_id_len, &lldp_peer_info->chassis_id);
+            lldp_peer_info->chassis_id[lldp_peer_info->chassis_id_len] = '\0';
+         }
+         else
+         {
+            LOG_ERROR(PF_LLDP_LOG, "LLDP(%d): Chassis Id too long (%d)\n", __LINE__, frame.len - 1);
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:580]: OK, I think I have reserved a enough buffer space for the chassis id. Max len for tlv format however is 512 bytes. If this is a problem we or customer will have to reconfigure PNET_LLDP_CHASSIS_ID_MAX_LEN and rebuild stack. Added comment.

offset variable increased.

File: src/common/pf_lldp.c (hunk 14 above)
Comment (olbjo) [reply src/common/pf_lldp.c:590]: ok

File: src/common/pf_lldp.c [hunk 21]
```diff
@@ -79,13 +100,48 @@ static const pnet_ethaddr_t   lldp_dst_addr = {
...
+ * - Length = 9 bits
+ * - data   = 0-511 bytes
+ *
+ * @param p_info           InOut: Parser state
+ * @param p_pos            InOut: Offset in parsed buffer
+ * @return parsed frame. If parsing fails, frame type is set to LLDP_TYPE_END and len field set to 0.
+ */
+lldp_frame_t pf_get_lldp_tlv(
+   pf_get_info_t           *p_info,
+   uint16_t                *p_pos)
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:127]: I think we should wait this update.  Also the buffer pointer in the pf_get_info_t struct could be const

File: src/common/pf_lldp.c (hunk 15 above)
Comment (olbjo) [reply src/common/pf_lldp.c:594]: ok

File: src/common/pf_lldp.c (hunk 16 above)
Comment (olbjo) [reply src/common/pf_lldp.c:642]: yes it should!
ok

File: src/common/pf_lldp.c [hunk 22]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
    }
 }
+
+/**
+ * @internal
+ * Parse LLDP frame buffer
+ *
+ * @param buf              In:    LLDP data.
+ * @param len              In:    Length of LLDP data.
+ * @param lldp_peer_info   InOut: Record holding parsed LLDP information.
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:544]: ok

File: include/pnet_api.h [hunk 23]
```diff
@@ -1027,17 +1035,78 @@ typedef struct pnet_ethaddr
...
+{
+   /* LLDP TVL's */
```
[the remark this answers is not in the document]
Comment (olbjo) [reply include/pnet_api.h:1089]: OK

File: src/common/pf_lldp.c [hunk 24]
```diff
@@ -54,23 +72,26 @@ typedef enum lldp_pnio_subtype_values
...
    { 0x01, 0x80, 0xc2, 0x00, 0x00, 0x0e }       /* LLDP Multicast */
 };
 
-
 /******************* Insert data into buffer ********************************/
 
 /**
  * @internal
  * Insert header of a TLV field into a buffer.
+ * Each TVL is structured as follows:
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:91]: OK

File: src/common/pf_lldp.c [hunk 25]
```diff
@@ -79,13 +100,48 @@ static const pnet_ethaddr_t   lldp_dst_addr = {
...
 {
-   pf_put_uint16(true, ((typ) << 9) + ((len) & 0x1ff), PF_FRAME_BUFFER_SIZE, p_buf, p_pos);
+   uint16_t tlv_header = ((typ) << LLDP_TYPE_SHIFT) + ((len) & LLDP_LENGTH_MASK);
+   pf_put_uint16(true, tlv_header, PF_FRAME_BUFFER_SIZE, p_buf, p_pos);
+}
+
+/**
+ * @internal
+ * Parse LLDP TLV header.
+ * Each TVL is structured as follows:
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:116]: OK

File: src/common/pf_lldp.c [hunk 26]
```diff
@@ -39,13 +42,28 @@
...
+#define LLDP_TYPE_SHIFT                   0x9
+#define LLDP_LENGTH_MASK                  0x1FF
+#define LLDP_MAX_TLV                      512
+
+typedef struct lldp_frame
+{
+   uint8_t  type;
+   uint16_t len;
+   uint8_t  *p_data;
+} lldp_frame_t;
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:60]: ok

File: src/common/pf_lldp.c [hunk 27]
```diff
@@ -79,13 +100,48 @@ static const pnet_ethaddr_t   lldp_dst_addr = {
...
+ *
+ * @param p_info           InOut: Parser state
+ * @param p_pos            InOut: Offset in parsed buffer
+ * @return parsed frame. If parsing fails, frame type is set to LLDP_TYPE_END and len field set to 0.
+ */
+lldp_frame_t pf_get_lldp_tlv(
+   pf_get_info_t           *p_info,
+   uint16_t                *p_pos)
+{
+   lldp_frame_t frame;
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:129]: ok

File: src/common/pf_lldp.c [hunk 28]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+ * @param len              In:    Length of LLDP data.
+ * @param lldp_peer_info   InOut: Record holding parsed LLDP information.
+ * @return 0 if parsing is successful, -1 on error.
+ */
+int pf_lldp_parse_packet(
+      uint8_t               buf[],
+      uint16_t              len,
+      pnet_lldp_peer_info_t *lldp_peer_info)
+{
+   lldp_frame_t         frame = {0};
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:552]: ok

File: include/pnet_api.h [hunk 29]
```diff
@@ -1027,17 +1035,78 @@ typedef struct pnet_ethaddr
...
+   uint8_t                             chassis_id_subtype;
+   char                                chassis_id[PNET_LLDP_CHASSIS_ID_MAX_LEN+1];
+   size_t                              chassis_id_len;
+   uint8_t                             port_id_subtype;
+   char                                port_id[PNET_LLDP_PORT_ID_MAX_LEN+1];
+   size_t                              port_id_len;
+   uint16_t                            ttl;
+   /* PROFIBUS TVL's */
```
[the remark this answers is not in the document]
Comment (olbjo) [reply include/pnet_api.h:1097]: OK

File: src/common/pf_lldp.c [hunk 30]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+
+   while (frame.type != LLDP_TYPE_END)
+   {
+      LOG_DEBUG(PF_LLDP_LOG, "LLDP(%d): lldp tlv type=%d len=%d\n", __LINE__, frame.type, frame.len);
+
+      switch (frame.type)
+      {
+      case LLDP_TYPE_CHASSIS_ID:
+         if ((frame.len > 0) &&
+             ((size_t)(frame.len - 1) < sizeof(lldp_peer_info->chassis_id)))
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:571]: I think original version is correct. Note that the tlv includes a subtype byte as well.

File: src/common/pf_lldp.c [hunk 31]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+         {
+            lldp_peer_info->chassis_id_len = frame.len - 1;
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:573]: I think original version is correct. Note that the tlv includes a subtype byte as well.

File: src/common/pf_lldp.c [hunk 32]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+            lldp_peer_info->chassis_id[lldp_peer_info->chassis_id_len] = '\0';
+         }
+         else
+         {
+            LOG_ERROR(PF_LLDP_LOG, "LLDP(%d): Chassis Id too long (%d)\n", __LINE__, frame.len - 1);
+         }
+         break;
+      case LLDP_TYPE_PORT_ID:
+         if ((frame.len > 0) &&
+             ((size_t)(frame.len - 1) < sizeof(lldp_peer_info->port_id)))
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:585]: I think original version is correct. Note that the tlv includes a subtype byte as well.

File: src/common/pf_lldp.c [hunk 33]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+         {
+            lldp_peer_info->port_id_len = frame.len - 1;
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:587]: I think original version is correct. Note that the tlv includes a subtype byte as well.

File: src/common/pf_lldp.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:548]: Other reused functions used for parsing needs to be updated if I change this so I would prefer to skip this. Is it ok to skip update?

File: src/common/pf_lldp.c [hunk 34]
```diff
@@ -440,10 +525,239 @@ void pf_lldp_send(
...
+
+   error = pf_lldp_generate_alias_name(lldp_peer_info->port_id, lldp_peer_info->chassis_id, alias, sizeof(alias));
+   if (!error && (strcmp(alias, net->cmina_current_dcp_ase.alias_name) != 0))
+   {
+      LOG_INFO(PF_LLDP_LOG, "LLDP(%d): Updating alias name: %s -> %s\n",
+            __LINE__, net->cmina_current_dcp_ase.alias_name,
+            alias);
+
+      memset(net->cmina_current_dcp_ase.alias_name, 0, sizeof(net->cmina_current_dcp_ase.alias_name));
+      strncpy(net->cmina_current_dcp_ase.alias_name, alias, sizeof(net->cmina_current_dcp_ase.alias_name));
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:734]: The alias buffer was smaller so I do not think it was a real problem. I have made an update so sizeof dcp_ase.alias_name is reused for the alias buffer.

File: src/common/pf_lldp.c (hunk 30 above)
[the remark this answers is not in the document]
Comment (fredrikdanielmoller) [reply src/common/pf_lldp.c]: My mistake. I missed the subtype byte.

File: src/common/pf_lldp.c (hunk 31 above)
[the remark this answers is not in the document]
Comment (fredrikdanielmoller) [reply src/common/pf_lldp.c]: My mistake. I missed the subtype byte.

File: src/common/pf_lldp.c (hunk 32 above)
[the remark this answers is not in the document]
Comment (fredrikdanielmoller) [reply src/common/pf_lldp.c]: My mistake. I missed the subtype byte.

File: src/common/pf_lldp.c (hunk 33 above)
[the remark this answers is not in the document]
Comment (fredrikdanielmoller) [reply src/common/pf_lldp.c]: My mistake. I missed the subtype byte.

File: src/common/pf_lldp.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (fredrikdanielmoller) [reply src/common/pf_lldp.c:548]: It's ok to skip. A separate pull request could add the 'const'.

File: src/common/pf_lldp.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (fredrikdanielmoller) [reply src/common/pf_lldp.c:127]: Ok

**Comment by olbjo** [comment]
Hopefully I have updated according to your comments. Please check if you find updates ok.

File: test/test_lldp.cpp [hunk 35]
```diff
@@ -59,3 +59,37 @@ TEST_F(LldpTest, LldpInitTest)
...
+   err =  pf_lldp_generate_alias_name("port-001", NULL, alias , 96);
+   EXPECT_EQ(err, -1);
+
+   err =  pf_lldp_generate_alias_name("port-001", "dut", NULL , 96);
+   EXPECT_EQ(err, -1);
+
+
+   err =  pf_lldp_generate_alias_name("port-001", "dut", alias, 13);
+   EXPECT_EQ(err, 0);
+   printf("alias :%s\n",alias);
```
Comment (hefloryd) [inline test/test_lldp.cpp:80]: Unit tests should preferably not clutter the output. See the TEST_TRACE macro if the output is useful for debugging the tests.

Comment (olbjo) [reply test/test_lldp.cpp:80]: Good point, the printf was left by mistake. I have pushed a version where it is removed

**Review by hefloryd — APPROVED** [review]

File: test/test_lldp.cpp (hunk 35 above)
Comment (fredrikdanielmoller) [reply test/test_lldp.cpp:80]: The 'printf' does not look removed. Am I looking at the wrong version?

Comment (olbjo) [reply test/test_lldp.cpp]: No, you are correct. I failed to remove the printf in the version i pushed. I however pushed a new update which fixes this.

File: test/test_lldp.cpp [hunk 36]
```diff
@@ -59,3 +59,36 @@ TEST_F(LldpTest, LldpInitTest)
...
+
+
+   err =  pf_lldp_generate_alias_name("port-001", "dut", alias, 13);
+   EXPECT_EQ(err, 0);
+   EXPECT_EQ(strcmp(alias, "port-001.dut"), 0);
+   memset(alias, 0, sizeof(alias));
+
+   err =  pf_lldp_generate_alias_name("port-001", "dut", alias, 12);
+   EXPECT_EQ(err, -1);
+
```
Comment (fredrikdanielmoller) [inline test/test_lldp.cpp:85]: Add tests for Profinet 2.2 legacy devices as well as Profinet 2.3+ devices.

**Review by fredrikdanielmoller — CHANGES_REQUESTED** [review]

File: src/common/pf_lldp.c [hunk 37]
```diff
@@ -143,33 +201,29 @@ static inline void pf_lldp_ieee_header(
...
    }
    else
    {
       /* Use the chassis_id from the cfg */
-      pf_lldp_tlv_header(p_buf, p_pos, LLDP_TYPE_CHASSIS_ID, 1+len);
+      pf_put_lldp_tlv_header(p_buf, p_pos, LLDP_TYPE_CHASSIS_ID, 1+len);
 
-      pf_put_byte(LLDP_SUBTYPE_CHASSIS_ID_NAME, PF_FRAME_BUFFER_SIZE, p_buf, p_pos);
-      pf_put_mem(p_station_name, len, PF_FRAME_BUFFER_SIZE, p_buf, p_pos);
+      pf_put_byte(LLDP_SUBTYPE_MAC, PF_FRAME_BUFFER_SIZE, p_buf, p_pos);
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:225]: OK

File: src/common/pf_lldp.c [hunk 38]
```diff
@@ -440,10 +541,246 @@ void pf_lldp_send(
...
+ * @param len              In:    Length of LLDP data.
+ * @param lldp_peer_info   Out:   Record holding parsed LLDP information.
+ * @return 0 if parsing is successful, -1 on error.
+ */
+int pf_lldp_parse_packet(
+      uint8_t               buf[],
+      uint16_t              len,
+      pnet_lldp_peer_info_t *lldp_peer_info)
+{
+   lldp_tlv_t           tlv = {0};
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:568]: OK

File: src/common/pf_lldp.c [hunk 39]
```diff
@@ -440,10 +541,246 @@ void pf_lldp_send(
...
+         {
+            lldp_peer_info->chassis_id_len = tlv.len - 1;
+            lldp_peer_info->chassis_id_subtype = pf_get_byte(&parse_info, &offset);
+            pf_get_mem(&parse_info, &offset, lldp_peer_info->chassis_id_len, &lldp_peer_info->chassis_id);
+            lldp_peer_info->chassis_id[lldp_peer_info->chassis_id_len] = '\0';
+         }
+         else
+         {
+            offset += tlv.len;
+            LOG_ERROR(PF_LLDP_LOG, "LLDP(%d): Chassis Id too long (%d)\n", __LINE__, tlv.len - 1);
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:597]: OK, I have added separate log message for tlv len <=1

File: src/common/pf_lldp.c [hunk 40]
```diff
@@ -440,10 +541,246 @@ void pf_lldp_send(
...
+         {
+            lldp_peer_info->port_id_len = tlv.len - 1;
+            lldp_peer_info->port_id_subtype = pf_get_byte(&parse_info, &offset);
+            pf_get_mem(&parse_info, &offset, lldp_peer_info->port_id_len, &lldp_peer_info->port_id);
+            lldp_peer_info->port_id[lldp_peer_info->port_id_len] = '\0';
+         }
+         else
+         {
+            offset += tlv.len;
+            LOG_ERROR(PF_LLDP_LOG, "LLDP(%d): Port Id too long (%d)\n", __LINE__, tlv.len - 1);
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:613]: OK, I have added separate log message for tlv len <=1

File: src/common/pf_lldp.c [hunk 41]
```diff
@@ -440,10 +541,246 @@ void pf_lldp_send(
...
+   const char*             chassis_id,
+   char*                   alias,
+   uint16_t                len)
+{
+   if (port_id == NULL || chassis_id == NULL || alias == NULL)
+   {
+      return -1;
+   }
+
+   if (strchr(port_id,'.') != NULL)
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:710]: ok

File: src/common/pf_lldp.c [hunk 42]
```diff
@@ -440,10 +541,246 @@ void pf_lldp_send(
...
+   {
+      if (strlen(port_id) >= len)
+      {
+         return -1;
+      }
+      strncpy(alias, port_id, len);
+   }
+   else
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:718]: ok

File: src/common/pf_lldp.c [hunk 43]
```diff
@@ -440,10 +541,246 @@ void pf_lldp_send(
...
+         return -1;
+      }
+      snprintf(alias, len, "%s.%s", port_id, chassis_id);
+   }
+   return 0;
+}
+
+/**
+ * Apply updated peer information
+ * @param net               In: p-net stack instance
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:731]: OK

File: src/common/pf_lldp.c [hunk 44]
```diff
@@ -440,10 +541,246 @@ void pf_lldp_send(
...
+         {
+            offset += tlv.len;
+            LOG_ERROR(PF_LLDP_LOG, "LLDP(%d): Port Id too long (%d)\n", __LINE__, tlv.len - 1);
+            /* Should not happen. Reconfigure PNET_LLDP_PORT_ID_MAX_LEN if it does */
+         }
+         break;
+      case LLDP_TYPE_TTL:
+         lldp_peer_info->ttl = pf_get_uint16(&parse_info, &offset);
+         break;
+      case LLDP_TYPE_ORG_SPEC:
```
[the remark this answers is not in the document]
Comment (olbjo) [reply src/common/pf_lldp.c:620]: OK, I have added a check of expected TLV length. Unexpected length will result in parsing error and warning message

File: test/test_lldp.cpp (hunk 36 above)
Comment (olbjo) [reply test/test_lldp.cpp:115]: I have added some comments and a test based on test data from rt-tester

**Review by fredrikdanielmoller — APPROVED** [review]

**Pull request merged**